src/Bundler.js

```diff
@@ -366,40 +366,34 @@ class Bundler extends EventEmitter {
     try {
       return await this.resolveAsset(dep.name, asset.name);
     } catch (err) {
-      let thrown = err;
-
-      if (thrown.message.indexOf(`Cannot find module '${dep.name}'`) === 0) {
-        // Check if dependency is a local file
-        let isLocalFile = /^[/~.]/.test(dep.name);
-        let fromNodeModules = asset.name.includes(
-          `${Path.sep}node_modules${Path.sep}`
-        );
-
-        // If it's not a local file, attempt to install the dep
-        if (
-          !isLocalFile &&
-          !fromNodeModules &&
-          this.options.autoinstall &&
-          install
-        ) {
-          return await this.installDep(asset, dep);
-        }
+      // If the dep is optional, return before we throw
+      if (dep.optional) {
+        return;
+      }
 
-        // If the dep is optional, return before we throw
-        if (dep.optional) {
-          return;
+      if (err.code === 'NODE_MODULE_NOT_FOUND' && install) {
+        let resolved;
+        if (dep.name.indexOf('/') > -1) {
+          // Check if module exists, prevents infinite install loop
+          try {
+            resolved = await this.resolver.resolve(
```

Comment:
You could move this call to `installDep` and return early if it fails.

One potential issue with trying to check if the package exists this way is with packages that don't have a `package.main` set. For example, imagine a package where there is no main entry point, but you instead need to request a sub-module every time e.g. `require('package')` doesn't work but `require('package/module')` does. It's not required to have a main entry point in a module. In this case, when someone did `require('package/bad-module')` where `package` is already installed but `package/bad-module` doesn't exist, we'd still get into a install loop.

Instead, we could just check if `node_modules/${this.resolver.getModuleParts(dep.name)[0]}` exists in any node_modules folder up to the root. In fact, we already do that check here: https://github.com/parcel-bundler/parcel/blob/master/src/Resolver.js#L149. In this case, we just don't want the code below which actually loads the package. Maybe we could extract the code that finds the node_modules folder containing the package from the rest of the code so we can reuse that function here. What do you think?

Reply:
I've changed it into your suggestions, however it will still install when it's a failed alias. Not sure how to properly handle that